dataclasses_avroschema/fields/fields.py

```diff
@@ -324,6 +325,116 @@ def fake(self) -> typing.Any:
         return field.fake()
 
 
+@dataclasses.dataclass
+class LiteralField(Field):
+    """
+    Supports typing.Literal. Note that typing.Literal[v1, v2, v3] is
+    Python's shorthand for typing.Union[typing.Literal[v1], typing.Literal[v2], typing.Literal[v3]]
+    and will be treated as such.
+
+    Note the following behavior when a string and matching enum are both
+    included in Literal:
+        class MyEnum(enum.Enum):
+            ONE = "one"
+
+        @dataclasses.dataclass
+        class MyModel(AvroModel):
+            field: typing.Literal["one", MyEnum.ONE]
+
+        MyModel.parse_obj({"field": "one"})
+
+    will always return:
+
+        # MyModel(field=<MyEnum.ONE: 'one'>)
+
+    This is because typing.Literal["one", MyEnum.ONE] is translated to
+    typing.Union[typing.Literal["one"], typing.Literal[MyEnum.ONE]]
+    and for matching a value to a union of literals, enums are prioritized
+    over other types.
+    """
+
+    allowed_values: typing.Set[typing.Any] = dataclasses.field(default_factory=set)
+    # This isn't actually optional, but Field has optional fields, so we have to provide a default
+    avro_field: typing.Optional[Field] = None
+
+    def __post_init__(self) -> None:
+        """
+        Derives Avro schema type[s] and validation requirements
+        """
+        args = get_args(self.type)
+        args_length = len(args)
+        if args_length > 1:
+            # This field is of the form typing.Literal[v1, v2, v3], which is a union of Literals
+            native_type = typing.Union[tuple(typing.Literal[a] for a in args)]  # type: ignore
+        else:
+            native_type = type(args[0])
+
+        self.allowed_values = set(args)
+        self.avro_field = AvroField(
+            name=self.name,
+            native_type=native_type,
+            parent=self.parent,
+            model_metadata=self.model_metadata,
+        )
+
+    def get_avro_type(self) -> typing.Any:
+        return self.avro_field.get_avro_type()  # type: ignore
+
+    def validate_value(self, value: typing.Any) -> bool:
+        """
+        Validates values assigned to Literal fields
+        """
+        if value not in self.allowed_values:
```

Comment:
Here we should only validate that the `default` matches the first `type` in the `Literal` (as we do with `Unions`), no business logic should be checked.

Now the following code is incorrect:

```python
import typing
import dataclasses

from dataclasses_avroschema import AvroModel

@dataclasses.dataclass
class MyModel(AvroModel):
    field: typing.Literal["one", 2] = 2


print(MyModel.print(MyModel.avro_schema())

{
  "type": "record", 
  "name": "MyModel", 
  "fields": [
    {"name": "field", "type": ["string", "long"],  "default": 2} # THIS IS WRONG!!
  ]
}
```

Reply:
I found that this use case it not working as well:

```python
import typing
from dataclasses import dataclass
import enum
from dataclasses_avroschema import AvroModel


class Versions(enum.Enum):
    v1 = "v1"
    v2 = "v2"


@dataclass
class Model(AvroModel):
    version: typing.Literal[Versions.v1, "one"] = Versions.v1

print(Model.avro_schema())

...
raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type Versions is not JSON serializable
```

Reply:
Is that with the latest commit? It works locally for me with the latest version of this branch [905fabc](https://github.com/marcosschroh/dataclasses-avroschema/pull/435/commits/905fabca65fed736b943eb9e7806fd4b363d61dd). I made a few fixes in the previous commit to properly support defaults.

Reply:
Maybe I did not pull the latest changes